purldb-toolkit/src/purldb_toolkit/purlcli.py

```diff
@@ -0,0 +1,254 @@
+
+
+@purlcli.command(name="versions")
+@click.option(
+    "--purl",
+    "purls",
+    multiple=True,
+    required=False,
+    help="PackageURL or PURL.",
+)
+@click.option(
+    "--output",
+    type=click.File(mode="w", encoding="utf-8"),
+    required=True,
+    default="-",
+    help="Write versions output as JSON to FILE.",
+)
+@click.option(
+    "--file",
+    type=click.File(mode="r", encoding="utf-8"),
+    required=False,
+    help="Read a list of PURLs from a FILE, one per line.",
+)
+def get_versions(purls, output, file):
+    """
+    Given one or more PURLs, return a list of all known versions for each PURL.
+    """
+    if purls and file:
+        raise click.UsageError("Use either purls or file but not both.")
+
+    if not (purls or file):
+        raise click.UsageError("Use either purls or file.")
+
+    if file:
+        purls = file.read().splitlines(False)
+
+    purl_versions = list_versions(purls)
+    json.dump(purl_versions, output, indent=4)
+
+
+def list_versions(purls):
+    purl_versions = []
+    for purl in purls:
+        dict_output = {}
+        dict_output["purl"] = purl
+        dict_output["versions"] = []
+
+        purl = purl.strip()
+        if not purl:
+            continue
+
+        if check_versions_purl(purl):
+            print(check_versions_purl(purl))
+            continue
+
+        for package_version_object in list(versions(purl)):
+            nested_dict = {}
+            nested_purl = purl + "@" + f'{package_version_object.to_dict()["value"]}'
+
+            nested_dict["purl"] = nested_purl
+            nested_dict["version"] = f'{package_version_object.to_dict()["value"]}'
+            nested_dict[
+                "release_date"
+            ] = f'{package_version_object.to_dict()["release_date"]}'
+
+            dict_output["versions"].append(nested_dict)
+
+        purl_versions.append(dict_output)
+
+    return purl_versions
+
+
+@purlcli.command(name="meta")
+@click.option(
+    "--purl",
+    "purls",
+    multiple=True,
+    required=False,
+    help="PackageURL or PURL.",
+)
+@click.option(
+    "--output",
+    type=click.File(mode="w", encoding="utf-8"),
+    required=True,
+    default="-",
+    help="Write versions output as JSON to FILE.",
+)
+@click.option(
+    "--file",
+    type=click.File(mode="r", encoding="utf-8"),
+    required=False,
+    help="Read a list of PURLs from a FILE, one per line.",
+)
+def get_meta(purls, output, file):
+    """
+    Given one or more PURLs, return a mapping of metadata fetched from the API for each PURL.
+    """
+    if purls and file:
+        raise click.UsageError("Use either purls or file but not both.")
+
+    if not (purls or file):
+        raise click.UsageError("Use either purls or file.")
+
+    if file:
+        purls = file.read().splitlines(False)
+
+    meta_info = get_meta_details(purls)
+    json.dump(meta_info, output, indent=4)
+
+
+def get_meta_details(purls):
+    meta_details = []
+    purl_versions = []
```

Comment:
Did you intend to use `meta_details` here, instead of `purl_versions`?

Reply:
Yes, I did intend to replace `purl_versions` with `meta_details` and finally did so just now.  Another good catch -- thank you.